### PR TITLE
kubeadm: Replace GetCoreImage with less error prone functions

### LINKED ISSUE #800: Investigate how kubeadm figures out versions of dependencies for a given version of k8s
## Is this a BUG REPORT or FEATURE REQUEST?
FEATURE REQUEST // REFACTOR

The way we figure out the version of etcd given a specific version of kubernetes is different (or at least in a different place) than the DNS versions (coredns/kube-dns). We will hit the same problem with the pause container.

There seems to be a notion of a core image. The fix might be considering the pause container as a core image and adding it to the [`GetCoreImage` function][getcoreimage]. 

[getcoreimage]: https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/images/images.go#L31

### PR DESCRIPTION
**Pull request by rosti**
**What this PR does / why we need it**:

GetCoreImage is a too generic function, that takes too many arguments. This
makes it prone to errors that may be difficult to trace. The solution is to
split it into the following couple of functions with a more targeted interface:

- GetKubeControlPlaneImage used to fetch Kubernetes control plane images or the
unified control plane image (if one is specified).

- GetEtcdImage is used to fetch the etcd image.

In addition to these, a couple of new utility functions are also created:

- GetKubeControlPlaneImageNoOverride used like GetKubeControlPlaneImage but
does not return the unified control plane image (even if it is set).

- GetGenericArchImage returns image path in the form of "prefix/image-goarch:tag"

Signed-off-by: Rostislav M. Georgiev <rostislavg@vmware.com>

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes kubernetes/kubeadm#800

**Special notes for your reviewer**:
/cc @kubernetes/sig-cluster-lifecycle-pr-reviews
/area kubeadm
/assign @luxas
/assign @timothysc
/cc @chuckha 
/kind enhancement

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Comment by timothysc** [comment]
this is out of scope for 1.11 and won't be reviewed for a while.

**Comment by luxas** [comment]
Additionally, I think we're gonna restructure the image handling in kubeadm wrt the config API types.
Probably giving each component and own image field...

**Comment by rosti** [comment]
Giving each control plane component its own image field can become prone to an error. I would suggest, that the K8s control plane components (and kube-proxy) be specified by a single setting in the config file.
The approach in this PR hides the complexity of the old GetCoreImage parameters. It, instead, introduces two functions, that will fetch the images based on the config settings (for the Kube control plane and etcd currently). If we want more flexibility on those lines, we can do something along the lines of:

```Go
type ComponentVersion struct {
        prefix, tag string // tag can be version though
}
```
And then use this to specify the prefix and tag for the control plane images (+ kube-proxy), etcd, kube-dns/CoreDNS and pause. In that case this PR is a step in the right direction and I can rework it on those lines.

@luxas @timothysc WDYT?

**Comment by luxas** [comment]
@rosti let's focus on how to handle images generally in the v1beta1 API first, and then we can come back to this and conclude whether this is a good implementation or not

File: cmd/kubeadm/app/images/images.go
Comment (luxas) [inline cmd/kubeadm/app/images/images.go]: Should be using the repoprefix as well, i.e. `k8s.gcr.io/coredns`

**Review by luxas — APPROVED** [review]
/approve
Ok, after some discussions, let's merge this after fixing the nit, and use this as a baseline for better image management in kubeadm going forward, with the config changes as well.

**Review by luxas — APPROVED** [review]
/lgtm

**Comment by luxas** [comment]
Thanks for this PR!

**Pull request merged**
